### PR TITLE
add support for multiple service annotations

### PR DESCRIPTION
**Pull request by wodka**
this only adds the ability to define multipe service annotations on a class

to fully implement https://github.com/schmittjoh/JMSDiExtraBundle/issues/232 we need to define how to add injections based on different service notations.

### CODE REVIEW

**Comment by wodka** [comment]
perhaps something like:

```
@Service(
  "first.service", 
  initMethod="setup", 
  initMethodInject=@InjectParams({
    "system" =@Inject("%first%)
  })
)
@Service(
  "second.service", 
  initMethod = "setup", 
  initMethodInject = @InjectParams({
    "system" = @Inject("%second%)
  })
)
```

Ideas?

File: Metadata/ClassMetadata.php [hunk 1]
```diff
@@ -34,6 +34,47 @@ class ClassMetadata extends BaseClassMetadata
...
+    public $services = [];
+
+    /**
+     * on first call also populate legacy fields
+     *
+     * @param string[] $service
+     */
+    public function addService(array $service)
+    {
+        if (empty($this->id)) {
```
Comment (Ener-Getick) [inline Metadata/ClassMetadata.php]: IMO we should deprecate all this attributes

File: Metadata/MetadataConverter.php [hunk 2]
```diff
@@ -37,49 +37,56 @@ public function convert(ClassHierarchyMetadata $metadata)
...
-                $definition->setPublic($classMetadata->public);
-            }
-            if (null !== $classMetadata->abstract) {
-                $definition->setAbstract($classMetadata->abstract);
-            }
-            if (null !== $classMetadata->arguments) {
-                $definition->setArguments($classMetadata->arguments);
-            }
+                $definition->setClass($classMetadata->name);
+                if (null !== @$service['scope']) {
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: why do you use ``@`` ?

File: Metadata/MetadataConverter.php [hunk 3]
```diff
@@ -37,49 +37,56 @@ public function convert(ClassHierarchyMetadata $metadata)
...
-            $definition->setProperties($classMetadata->properties);
+                $definition->setMethodCalls($classMetadata->methodCalls);
+                $definition->setTags($classMetadata->tags);
+                $definition->setProperties($classMetadata->properties);
 
-            if (null === $classMetadata->id) {
-                $classMetadata->id = '_jms_di_extra.unnamed.service_'.$count++;
-            }
+                if (null === @$service['id']) {
+                    $classMetadata->id = '_jms_di_extra.unnamed.service_' . $count++;
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: It looks strange to me to use this ``ClassMetadata::$id`` as it is now deprecated...

File: Metadata/MetadataConverter.php [hunk 4]
```diff
@@ -37,49 +37,56 @@ public function convert(ClassHierarchyMetadata $metadata)
...
             }
 
-            $definitions[$classMetadata->id] = $definition;
             $previous = $classMetadata;
         }
 
         return $definitions;
     }
+
+    protected function buildServices(ClassMetadata $classMetadata, $previous)
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: What is that for ?

File: Metadata/MetadataConverter.php [hunk 5]
```diff
@@ -37,49 +37,56 @@ public function convert(ClassHierarchyMetadata $metadata)
...
+
+                if ($classMetadata->initMethod) {
+                    if (!method_exists($definition, 'setInitMethod')) {
+                        throw new \RuntimeException(sprintf('@AfterSetup is not available on your Symfony version.'));
+                    }
 
-            if ($classMetadata->initMethod) {
-                if (!method_exists($definition, 'setInitMethod')) {
-                    throw new \RuntimeException(sprintf('@AfterSetup is not available on your Symfony version.'));
+                    $definition->setInitMethod($classMetadata->initMethod);
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: This method doesn't exist. Any idea why this was introduce in first place ?

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: This feature doesn't seem to work, we should maybe remove it.

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: I changed this exception message, please rebase your PR on master.

File: Metadata/ClassMetadata.php [hunk 6]
```diff
@@ -34,6 +34,47 @@ class ClassMetadata extends BaseClassMetadata
     public $properties = array();
     public $initMethod;
     public $environments = array();
+    public $services = [];
```
Comment (Ener-Getick) [inline Metadata/ClassMetadata.php]: You must use the long syntax ``array()``

File: Tests/Metadata/Driver/AnnotationDriverTest.php [hunk 7]
```diff
@@ -43,6 +43,19 @@ public function testCustomAnnotationOnMethod()
...
+    public function testMultiService()
+    {
+        $metadata = $this->getDriver()->loadMetadataForClass(new \ReflectionClass('JMS\DiExtraBundle\Tests\Metadata\Driver\Fixture\MultiService'));
+
+        $this->assertEquals(array(
+            'first.service' => array(
+                'id' => 'first.service',
+            ),
+            'second.service' => array(
+                'id' => 'first.service',
```
Comment (Ener-Getick) [inline Tests/Metadata/Driver/AnnotationDriverTest.php]: should be ``second.service`` here.

Comment (Ener-Getick) [inline Tests/Metadata/Driver/AnnotationDriverTest.php]: BTW your array is not corresponding to what is tested:
```diff
 Array (
     'first.service' => Array (
         'id' => 'first.service'
+        'parent' => null
+        'public' => null
+        'scope' => null
+        'abstract' => null
     )
     'second.service' => Array (
-        'id' => 'first.service'
+        'id' => 'second.service'
+        'parent' => null
+        'public' => null
+        'scope' => null
+        'abstract' => null
     )
 )
```

File: Metadata/MetadataConverter.php (hunk 2 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: at first I only added values to the array if they existed in AnnotationDriver ~ line 82, I will change the checks now to !empty

File: Metadata/MetadataConverter.php (hunk 4 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: missed to delete it ...

File: Metadata/MetadataConverter.php (hunk 3 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: removed it now and changed the other occurrences to the service definition

File: Metadata/MetadataConverter.php (hunk 5 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: no clue where this comes from - this is not present in the symfony repository even in 2011 when it was added

I think it should be removed in another pull request

File: Metadata/ClassMetadata.php [hunk 8]
```diff
@@ -34,6 +34,55 @@ class ClassMetadata extends BaseClassMetadata
...
+     */
+    public function hasServices()
+    {
+        return !empty($this->services);
+    }
+
+    /**
+     * get list of defined services, use fallback of original fields
+     *
+     * return array[]
```
Comment (Ener-Getick) [inline Metadata/ClassMetadata.php]: ``@return``

File: Metadata/ClassMetadata.php [hunk 9]
```diff
@@ -34,6 +34,55 @@ class ClassMetadata extends BaseClassMetadata
     public $properties = array();
     public $initMethod;
     public $environments = array();
+    public $services = array();
```
Comment (Ener-Getick) [inline Metadata/ClassMetadata.php]: As there is a getter IMO this should be private

File: Metadata/Driver/AnnotationDriver.php [hunk 10]
```diff
@@ -72,30 +72,38 @@ public function loadMetadataForClass(\ReflectionClass $class)
...
-                $metadata->public = $annot->public;
-                $metadata->scope = $annot->scope;
-                $metadata->abstract = $annot->abstract;
+                $service['parent'] = $annot->parent;
+                $service['public'] = $annot->public;
+                $service['scope'] = $annot->scope;
+                $service['abstract'] = $annot->abstract;
+
+                // remove all empty values from service definition
+                $metadata->addService(array_filter($service));
```
Comment (Ener-Getick) [inline Metadata/Driver/AnnotationDriver.php]: IMO we should not remove empty values.

Comment (wodka) [inline Metadata/Driver/AnnotationDriver.php]: oki, then I add this to the unit test - else we would always have to update that one just because a new null value was added.

Comment (Ener-Getick) [inline Metadata/Driver/AnnotationDriver.php]: I think there are more benefits to keep them (we don't have to always check if the values exist).

File: Metadata/ClassMetadata.php (hunk 1 above)
Comment (wodka) [inline Metadata/ClassMetadata.php]: I added the deprecated annotation to old properties

File: Metadata/MetadataConverter.php [hunk 11]
```diff
@@ -36,43 +37,55 @@ public function convert(ClassHierarchyMetadata $metadata)
...
-            } else {
-                $definition = new DefinitionDecorator(
-                    $classMetadata->parent ?: $previous->id
-                );
-            }
+            foreach ($classMetadata->getServices() as $service) {
+                if (null === $previous && empty($service['parent'])) {
+                    $definition = new Definition();
+                } else {
+                    if (empty($service['parent']) && sizeof($previous->getServices()) > 1) {
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: Please use ``null ===``` instead of ``empty``

Comment (wodka) [inline Metadata/MetadataConverter.php]: I can change it to !isset() - else I have to readd the @ (or copy&paste the null values for every addService call)

**Comment by wodka** [comment]
@Ener-Getick I just added the support for AfterSetup services

File: Metadata/MetadataConverter.php [hunk 12]
```diff
@@ -30,75 +30,123 @@ class MetadataConverter
...
         foreach ($metadata->classMetadata as $classMetadata) {
-            if (null === $previous && null === $classMetadata->parent) {
-                $definition = new Definition();
-            } else {
-                $definition = new DefinitionDecorator(
-                    $classMetadata->parent ?: $previous->id
-                );
-            }
+            foreach ($classMetadata->getServices() as $service) {
+                if (isset($environment)
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: always true

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: Use ``null !==`` instead, it is clearer

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php [hunk 13]
```diff
@@ -55,14 +55,12 @@ public function process(ContainerBuilder $container)
...
+
             if (null === $metadata->getOutsideClassMetadata()->id) {
                 continue;
             }
-            if ( ! $metadata->getOutsideClassMetadata()->isLoadedInEnvironment($container->getParameter('kernel.environment'))) {
-                continue;
-            }
 
-            foreach ($converter->convert($metadata) as $id => $definition) {
+            foreach ($converter->convert($metadata, $container->getParameter('kernel.environment')) as $id => $definition) {
```
Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: Is this related?

File: Metadata/MetadataConverter.php (hunk 12 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: changed it

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php (hunk 13 above)
Comment (wodka) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: before the environment check could be done on class level, but since we have now multiple services per class we have to check in there

**Comment by Ener-Getick** [comment]
I'm wondering if we should not just create a new ``ServiceMetadata`` class or just use the ``Definition`` class directly...

**Comment by Ener-Getick** [comment]
@schmittjoh What do you think about having a new metadata class ? something like ``ServiceMetadata`` which would contain all the linked fields instead of having an array.

**Comment by Ener-Getick** [comment]
as this is a big stuff, I think it will wait for 1.8.
